compiler/rustc_trait_selection/src/traits/specialize/mod.rs

```diff
@@ -595,3 +598,128 @@ fn report_conflicting_impls<'tcx>(
         }
     }
 }
+
+pub(super) fn trait_has_impl_which_may_shadow_dyn<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    (target_trait_def_id, principal_def_id): (DefId, Option<DefId>),
+) -> bool {
+    // We only care about trait objects which have associated types.
+    if !tcx
+        .associated_items(target_trait_def_id)
+        .in_definition_order()
+        .any(|item| item.kind == ty::AssocKind::Type)
+    {
+        return false;
+    }
+
+    let target_self_ty =
+        principal_def_id.map_or(SimplifiedType::MarkerTraitObject, SimplifiedType::Trait);
```

Comment:
```suggestion
        SimplifiedType::Trait(principal_def_id.unwrap());
```
do we get here with auto-traits? I don't think so, as we'd always return `true`  in this case